sys/net/application_layer/cord/lc/cord_lc.c

```diff
@@ -0,0 +1,432 @@
+static volatile thread_t *_waiter;
+
+static void _lock(void)
+{
+    mutex_lock(&_mutex);
+    _waiter = sched_active_thread;
+}
+
+static int _sync(void)
+{
+    thread_flags_t flags = thread_flags_wait_any(FLAG_MASK);
+
+    if (flags & FLAG_ERR) {
+        return CORD_LC_ERR;
+    } else if (flags & FLAG_TIMEOUT) {
+        return CORD_LC_TIMEOUT;
+    } else if (flags & FLAG_OVERFLOW) {
+        return CORD_LC_OVERFLOW;
+    } else if (flags & FLAG_NORSC) {
+        return CORD_LC_NORSC;
+    } else {
+        return CORD_LC_OK;
+    }
+}
+
+static void _on_lookup(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
+                       const sock_udp_ep_t *remote)
+{
+    (void)remote;
+
+    thread_flags_t flag = FLAG_ERR;
+
+    if (memo->state == GCOAP_MEMO_RESP) {
+        unsigned ct = coap_get_content_type(pdu);
+        if (ct != COAP_FORMAT_LINK) {
+            DEBUG("cord_lc: unsupported content format\n");
+            goto end;
+        }
+        if (pdu->payload_len == 0) {
+            flag = FLAG_NORSC;
+            goto end;
+        }
+        if (pdu->payload_len >= _result_buf_len) {
+            flag = FLAG_OVERFLOW;
+            goto end;
+        }
+        memcpy(_result_buf, pdu->payload, pdu->payload_len);
+        memset(_result_buf + pdu->payload_len, 0,
+               _result_buf_len - pdu->payload_len);
+        _result_buf_len = pdu->payload_len;
+        flag = FLAG_SUCCESS;
+    } else if (memo->state == GCOAP_MEMO_TIMEOUT) {
+        flag = FLAG_TIMEOUT;
+    }
+
+end:
+    thread_flags_set((thread_t *)_waiter, flag);
+}
+
+static ssize_t _add_filters_to_lookup(coap_pkt_t *pkt, cord_lc_filter_t *filters)
+{
+    cord_lc_filter_t *f = filters;
+    while (f) {
+        for (unsigned i = 0; i < f->len; i++) {
+            char key[NANOCOAP_QS_MAX];
+            char val[NANOCOAP_QS_MAX];
+
+            clif_attr_t *kv = &(f->array[i]);
+            if ((kv->key_len + kv->value_len + 1) >= NANOCOAP_QS_MAX ) {
+                DEBUG("cord_lc: filter key value pair overflow\n");
+                return CORD_LC_OVERFLOW;
+            }
+            snprintf(key, kv->key_len + 1, "%s", kv->key);
+            snprintf(val, kv->value_len + 1, "%s", kv->value);
+
+            if (gcoap_add_qstring(pkt, key, val) < 0) {
+                return CORD_LC_ERR;
+            }
+        }
+        f = f->next;
+    }
+    return CORD_LC_OK;
+}
+
+static ssize_t _lookup_raw(const cord_lc_rd_t *rd, unsigned content_format,
+                           unsigned lookup_type, cord_lc_filter_t *filters,
+                           void *result, size_t maxlen)
+{
+    assert(rd->remote);
+
+    uint8_t buf[GCOAP_PDU_BUF_SIZE];
+    memset(buf, 0, sizeof(buf));
+
+    int res;
+    int retval;
+    coap_pkt_t pkt;
+    ssize_t pkt_len;
+
+    char lookif[NANOCOAP_QS_MAX];
+    memset(lookif, 0, sizeof(lookif));
+    switch (lookup_type) {
+    case CORD_LC_RES:
+        memcpy(lookif, rd->res_lookif, rd->res_lookif_len);
```

Comment:
why copy this value? A simple pointer should do the trick. Like this its just a waste of memory.

Reply:
I agree, though it requires `res_lookif` and `ep_lookif` to be null terminated strings, but I guess its worth 1 byte to save the length param and the `lookif` buffer here plus the `memcpy`, too.